File: ios/Runner/MetronomePlugin.swift
import Flutter
import UIKit

/// Flutter plugin registering MethodChannel + EventChannel for sample-accurate metronome.
///
/// MethodChannel 'com.metronome.app/metronome':
///   start(bpm, beatsPerBar, patternSlots, timbreId)
///   stop()
///   updateConfig(bpm, beatsPerBar, patternSlots)
///   setTimbre(timbreId)
///
/// EventChannel 'com.metronome.app/metronome/beats':
///   emits {beatIndex: Int, slotIndex: Int, slotType: Int} on each slot
class MetronomePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var metronome: SampleAccurateMetronome?
    private var eventSink: FlutterEventSink?

    // MARK: - Registration via FlutterPlugin protocol (used by FlutterPluginRegistry)

    static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = MetronomePlugin()
        plugin.setupChannels(messenger: registrar.messenger())
        registrar.publish(plugin)
    }

    // MARK: - Registration via direct messenger (used by applicationRegistrar)

    static func register(messenger: FlutterBinaryMessenger) {
        let plugin = MetronomePlugin()
        plugin.setupChannels(messenger: messenger)
    }

    // MARK: - Channel setup

    private func setupChannels(messenger: FlutterBinaryMessenger) {
        let methodChannel = FlutterMethodChannel(
            name: "com.metronome.app/metronome",
            binaryMessenger: messenger)
        let eventChannel = FlutterEventChannel(
            name: "com.metronome.app/metronome/beats",
            binaryMessenger: messenger)

        methodChannel.setMethodCallHandler(handle)
        eventChannel.setStreamHandler(self)
    }

    // MARK: - MethodChannel handler

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            guard let args = call.arguments as? [String: Any],
                  let bpm = args["bpm"] as? Int,
                  let beatsPerBar = args["beatsPerBar"] as? Int,
                  let patternSlots = args["patternSlots"] as? [Int] else {
                result(FlutterError(code: "BAD_ARGS", message: "start requires bpm/beatsPerBar/patternSlots", details: nil))
                return
            }
            let timbreId = args["timbreId"] as? String ?? "click"
            let m = SampleAccurateMetronome(bpm: bpm, beatsPerBar: beatsPerBar, patternSlots: patternSlots)
            m.onBeat = { [weak self] payload in self?.eventSink?(payload) }
            m.setTimbre(timbreId)
            metronome = m
            do {
                try m.start()
                result(nil)
            } catch {
                result(FlutterError(code: "START_FAILED", message: error.localizedDescription, details: nil))
            }

        case "stop":
            metronome?.stop()
            metronome = nil
            result(nil)

        case "updateConfig":
            guard let args = call.arguments as? [String: Any],
                  let bpm = args["bpm"] as? Int,
                  let beatsPerBar = args["beatsPerBar"] as? Int,
                  let patternSlots = args["patternSlots"] as? [Int] else {
                result(FlutterError(code: "BAD_ARGS", message: "updateConfig requires bpm/beatsPerBar/patternSlots", details: nil))
                return
            }
            metronome?.updateConfig(bpm: bpm, beatsPerBar: beatsPerBar, patternSlots: patternSlots)
            result(nil)

        case "setTimbre":
            guard let args = call.arguments as? [String: Any],
                  let timbreId = args["timbreId"] as? String else {
                result(FlutterError(code: "BAD_ARGS", message: "setTimbre requires timbreId", details: nil))
                return
            }
            metronome?.setTimbre(timbreId)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/metronome/metronome_app/MetronomePlugin.kt
package com.metronome.metronome_app

import android.content.Context
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter plugin bridging MethodChannel + EventChannel to AudioTrackMetronome.
 *
 * MethodChannel  'com.metronome.app/metronome': start, stop, updateConfig, setTimbre
 * EventChannel   'com.metronome.app/metronome/beats': beat maps → Dart UI
 */
class MetronomePlugin(private val context: Context) {

    companion object {
        private const val METHOD_CHANNEL = "com.metronome.app/metronome"
        private const val EVENT_CHANNEL = "com.metronome.app/metronome/beats"

        fun registerWith(engine: FlutterEngine, context: Context) {
            val plugin = MetronomePlugin(context)
            plugin.setup(engine)
        }
    }

    private var metronome: AudioTrackMetronome? = null
    private var eventSink: EventChannel.EventSink? = null

    fun setup(engine: FlutterEngine) {
        val messenger = engine.dartExecutor.binaryMessenger

        MethodChannel(messenger, METHOD_CHANNEL).setMethodCallHandler { call, result ->
            when (call.method) {
                "start" -> {
                    val args = call.arguments as? Map<*, *>
                    val bpm = args?.get("bpm") as? Int ?: 120
                    val beatsPerBar = args?.get("beatsPerBar") as? Int ?: 4
                    @Suppress("UNCHECKED_CAST")
                    val patternSlots = (args?.get("patternSlots") as? List<*>)
                        ?.filterIsInstance<Int>() ?: listOf(0, 1, 1, 1)
                    val timbreId = args?.get("timbreId") as? String ?: "click"
                    val m = AudioTrackMetronome(context)
                    m.onBeat = { payload ->
                        eventSink?.success(payload)
                    }
                    m.updateConfig(bpm, beatsPerBar, patternSlots)
                    m.setTimbre(timbreId)
                    metronome = m
                    m.start()
                    result.success(null)
                }
                "stop" -> {
                    metronome?.onBeat = null
                    metronome?.stop()
                    metronome = null
                    result.success(null)
                }
                "updateConfig" -> {
                    val args = call.arguments as? Map<*, *>
                    val bpm = args?.get("bpm") as? Int ?: 120
                    val beatsPerBar = args?.get("beatsPerBar") as? Int ?: 4
                    @Suppress("UNCHECKED_CAST")
                    val patternSlots = (args?.get("patternSlots") as? List<*>)
                        ?.filterIsInstance<Int>() ?: listOf(0, 1, 1, 1)
                    metronome?.updateConfig(bpm, beatsPerBar, patternSlots)
                    result.success(null)
                }
                "setTimbre" -> {
                    val timbreId = (call.arguments as? Map<*, *>)?.get("timbreId") as? String ?: "click"
                    metronome?.setTimbre(timbreId)
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }

        EventChannel(messenger, EVENT_CHANNEL).setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                eventSink = events
            }
            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }
}
